ash/src/vk.rs

```diff
@@ -82,6 +66,103 @@ pub(crate) unsafe fn ptr_chain_iter<T: ?Sized>(
+    /// `unsafe` [`Self::extend()`] to insert this chain instead.
+    fn push<T: Extends<'a, Self> + ?Sized>(mut self, next: &'a mut T) -> Self
+    where
+        Self: Sized,
+    {
+        // SAFETY: All implementors of `AnyTaggedStructure` are required to have the `BaseOutStructure` layout
+        let slf_base = unsafe { &mut *<*mut _>::cast::<BaseOutStructure<'_>>(&mut self) };
+        // SAFETY: All implementors of `T: Extends<'a, >: AnyTaggedStructure` are required to have the `BaseOutStructure` layout
+        let next_base = unsafe { &mut *<*mut T>::cast::<BaseOutStructure<'_>>(next) };
+        // `next` here can contain a pointer chain.  This function refuses to insert the struct,
+        // in favour of calling unsafe extend().
+        assert!(
+            next_base.p_next.is_null(),
+            "push() expects a struct without an existing p_next pointer chain (equal to NULL)"
+        );
+        next_base.p_next = slf_base.p_next;
+        slf_base.p_next = next_base;
+        self
+    }
+
+    /// Prepends the given extension struct between the root and the first pointer. This
+    /// method only exists on structs that can be passed to a function directly. Only
+    /// valid extension structs can be pushed into the chain.
+    /// If the chain looks like `A -> B -> C` and `D -> E`, and you call `A.extend(&mut D)`,
+    /// then the chain will look like `A -> D -> E -> B -> C`.
+    ///
+    /// # Safety
+    /// This function will walk the [`BaseOutStructure::p_next`] chain of `next`, requiring
+    /// all non-`NULL` pointers to point to a valid Vulkan structure starting with the
+    /// [`BaseOutStructure`] layout.
+    ///
+    /// The last struct in this chain (i.e. the one where `p_next` is `NULL`) must
+    /// be writable memory, as its `p_next` field will be updated with the value of
+    /// `self.p_next`.
+    unsafe fn extend<T: Extends<'a, Self> + ?Sized>(mut self, next: &'a mut T) -> Self
+    where
+        Self: Sized,
+    {
+        // `next` here can contain a pointer chain. This means that we must correctly
+        // attach he head to the root and the tail to the rest of the chain
+        // For example:
+        //
+        // next = A -> B
+        // Before: `Root -> C -> D -> E`
+        // After: `Root -> A -> B -> C -> D -> E`
+        //                 ^^^^^^
+        //                 next chain
+        let slf_base = unsafe { &mut *<*mut _>::cast::<BaseOutStructure<'_>>(&mut self) };
+        let next_base = <*mut T>::cast::<BaseOutStructure<'_>>(next);
+        let last_next = ptr_chain_iter(next).last().unwrap();
+        (*last_next).p_next = slf_base.p_next;
+        slf_base.p_next = next_base;
+        self
+    }
+
+    #[doc(hidden)]
+    #[deprecated = "Migrate to `push()` if `next` does not have an existing chain (i.e. `p_next` is `NULL`), `extend()` otherwise"]
+    unsafe fn push_next<T: Extends<'a, Self> + ?Sized>(self, next: &'a mut T) -> Self
```

Comment:
Should we take the opportunity to kill this?

Reply:
Since we're planning on a breaking release, I'm more than happy to not keep legacy deprecated code around; folks should read the changelog and learn how/why to migrate their original `.push_next()` to either call `push()` or `extend()` instead.